### DISCUSSION TITLE
Weasyprint Dependencies

### DISCUSSION THREAD

**Discussion by TRezendes** [beeware/briefcase#2054] [Q&A]
I have a python script that uses Weasyprint to generate PDFs. I am using [swoopyui](https://github.com/SKbarbon/swoopyui) for the interface and briefcase to package it. When I open the app after briefcase creates it, I get a window with "Application crashed. An unexpected error occurred. Please see the traceback below for more information."

```
Traceback (most recent call last):
  File "Bar Code Generator.app/Contents/Resources/app/barcodegenerator/__main__.py", line 1, in <module>
    from barcodegenerator.app import main
  File "Bar Code Generator.app/Contents/Resources/app/barcodegenerator/app.py", line 6, in <module>
    from .bcg import barcode_printer
  File "Bar Code Generator.app/Contents/Resources/app/barcodegenerator/bcg.py", line 7, in <module>
    from weasyprint import CSS, HTML
  File "Bar Code Generator.app/Contents/Resources/app_packages/weasyprint/__init__.py", line 419, in <module>
    from .css import preprocess_stylesheet  # noqa: I001, E402
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Bar Code Generator.app/Contents/Resources/app_packages/weasyprint/css/__init__.py", line 28, in <module>
    from .computed_values import COMPUTER_FUNCTIONS
  File "Bar Code Generator.app/Contents/Resources/app_packages/weasyprint/css/computed_values.py", line 9, in <module>
    from ..text.ffi import ffi, pango, units_to_double
  File "Bar Code Generator.app/Contents/Resources/app_packages/weasyprint/text/ffi.py", line 431, in <module>
    gobject = _dlopen(
              ^^^^^^^^
  File "Bar Code Generator.app/Contents/Resources/app_packages/weasyprint/text/ffi.py", line 420, in _dlopen
    return ffi.dlopen(names[0])  # pragma: no cover
           ^^^^^^^^^^^^^^^^^^^^
  File "Bar Code Generator.app/Contents/Resources/app_packages/cffi/api.py", line 150, in dlopen
    lib, function_cache = _make_ffi_library(self, name, flags)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Bar Code Generator.app/Contents/Resources/app_packages/cffi/api.py", line 834, in _make_ffi_library
    backendlib = _load_backend_lib(backend, libname, flags)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Bar Code Generator.app/Contents/Resources/app_packages/cffi/api.py", line 829, in _load_backend_lib
    raise OSError(msg)
OSError: cannot load library 'gobject-2.0-0': dlopen(gobject-2.0-0, 0x0002): tried: 'gobject-2.0-0' (no such file), '/System/Volumes/Preboot/Cryptexes/OSgobject-2.0-0' (no such file), '/Users/trezendes/Projects/LCDSbarcodes/barcodegenerator/build/barcodegenerator/macos/app/Bar Code Generator.app/Contents/Frameworks/gobject-2.0-0' (no such file), '/usr/lib/gobject-2.0-0' (no such file, not in dyld cache), 'gobject-2.0-0' (no such file).  Additionally, ctypes.util.find_library() did not manage to locate a library called 'gobject-2.0-0'
```

Below are the requirements.txt and the imports from the project.

requirements.txt
```
blinker==1.8.2
Brotli==1.1.0
certifi==2024.8.30
cffi==1.17.1
charset-normalizer==3.4.0
click==8.1.7
cssselect2==0.7.0
Flask==3.0.3
fonttools==4.54.1
html5lib==1.1
idna==3.10
itsdangerous==2.2.0
Jinja2==3.1.4
MarkupSafe==3.0.2
pillow==11.0.0
pycparser==2.22
pydyf==0.11.0
pyphen==0.16.0
python-barcode==0.15.1
requests==2.32.3
six==1.16.0
swoopyui==2.7
tinycss2==1.4.0
urllib3==2.2.3
weasyprint==62.3
webencodings==0.5.1
Werkzeug==3.0.6
zopfli==0.2.3.post1
```

app.py
```
import re
from time import sleep

from swoopyui import app, HStack, Icon, Spacer, Text, TextButton, TextField, View, VStack, ZStack

from .bcg import barcode_printer
```

bcg.py
```
import os
import base64
from io import BytesIO

from barcode import Code39
from barcode.writer import SVGWriter
from weasyprint import CSS, HTML
from weasyprint.text.fonts import FontConfiguration
```

I more or less expected Weasyprint to cause a problem, and I think I basically understand *why* it is a problem. My question is, is there a workaround? Is there a way to package this app so that it runs as an app? For my purposes, it works just as well if Weasyprint's dependencies are bundled in the app or if the app is able to utilize resources that are `brew install`ed independently.

I would be grateful for any insight on this matter.

**Comment by freakboy3742** [comment]
Based on the references to `brew` and the stack trace, I'm assuming you're packaging for macOS.

I haven't used SwoopyUI or weasyprint, so I can't comment with specifics. However, the stack trace suggests that weasyprint isn't a *complete* wheel - that is, it doesn't include all the binaries that it needs at runtime. Instead, it's assuming it can dynamically load a `gobject` DLL at runtime.  That isn't a use case that Briefcase currently supports out of the box - Briefcase assumes that the installed wheels will be self-contained. 

However, you might be able to make it work. There's essentially two possible approaches:
1. Modify the app so that it can load the DLL from homebrew. This essentially means either ensuring the homebrew lib directory is on your `DYLD_LIBRARY_PATH` when the app runs, or modifying the code so that you provide an *explicit* path to the homebrew library (i.e., a full path, not just "libgobject-2.0.dylib").
2. Manually add libgobject-2.0.dylib to your app bundle in a location where it will be included at runtime. This likely involves converting the dylib into a Framework - this is essentially the same library, but in a folder named `gobject-2.0.framework`, with an `Info.plist` file that describes the binary. You'll also need to delocate the binary so that there aren't any transitive dependencies on *other* libraries, and the library's name isn't hard-coded to include an absolute path.

I guess there's also a third option of "fix weasyprint so that its wheels are self-contained". That *might* end up being the easier approach.
